### DISCUSSION TITLE
Force Argo to "take ownership" of child resources to fail syncs on unhealthy resources

### DISCUSSION THREAD

**Discussion by daveoy** [argoproj/argo-cd#17199] [Q&A]
im attempting to use preSync and postSync hooks to control alert silencing while a deployment is rolling out. the preSync and postSync jobs are working well, and doing what they're meant to do; however, it seems i have an ordering issue or a fundamental misunderstanding im hoping the community can help resolve for me.

here is the cycle of events following a click of the sync button in argo during which i *know* the pods spawned by the deployment template in this argo app will fail to start (incorrect image tag for testing purposes)

```
»  kubectl get pods -w
NAME                                   READY   STATUS      RESTARTS   AGE
deploy-alert-silence-c6xgx             0/1     Completed   0          5s
node-cycle-analyzer-7b6cf59779-r24v5   1/1     Running     0          3d18h
node-cycle-analyzer-85f9467776-48zsg   0/1     Pending     0          0s
node-cycle-analyzer-85f9467776-48zsg   0/1     Pending     0          0s
node-cycle-analyzer-85f9467776-48zsg   0/1     ContainerCreating   0          0s
node-cycle-analyzer-85f9467776-48zsg   0/1     ContainerCreating   0          3s
node-cycle-analyzer-85f9467776-48zsg   0/1     ErrImagePull        0          4s
node-cycle-analyzer-85f9467776-48zsg   0/1     ImagePullBackOff    0          5s
remove-alert-silence-djh9j             0/1     Pending             0          0s
remove-alert-silence-djh9j             0/1     Pending             0          0s
remove-alert-silence-djh9j             0/1     ContainerCreating   0          0s
remove-alert-silence-djh9j             0/1     ContainerCreating   0          3s
remove-alert-silence-djh9j             0/1     Completed           0          4s
remove-alert-silence-djh9j             0/1     Completed           0          4s
```

the silence is deployed by the preSync hook, the deployment is updated (you can see the replicaset hash change), but even though the pod spawned by this deployment is in ImagePullBackOff, the postSync hook fires.

is it possible to have a sync stop and fail when a child resource of something in the syncing app's helm chart is unhealthy?

I have tried forcing argo to "take ownership" of the pods spawned by the deployment in this app by adding `argocd.argoproj.io/instance:` labels to the pod template, which properly assigned the labels to new pods, but made no behavioural change in argo.

any input is appreciated!

**Comment by daveoy** [comment]
also mentioned in slack https://cloud-native.slack.com/archives/C01TSERG0KZ/p1707849016940009

**Comment by jannfis** [comment]
Sync (and sync status) is fundamentally different from the concept of health, so the health of your resources won't affect sync status.

But to me, it seems like your use case could be solved using [sync-waves](https://argo-cd.readthedocs.io/en/stable/user-guide/sync-waves/#how-do-i-configure-waves). With multiple waves, your sync will only proceed if the current phase has completed, and completed in this sense means all resources are in Synced state and are healthy.

> I have tried forcing argo to "take ownership" of the pods spawned by the deployment in this app by adding argocd.argoproj.io/instance: labels to the pod template

Don't do that. In case of Pods, as they have ownerReferences pointing back to a deployment, Argo will just ignore them. But if you do that to other resources, you might get side-effects leading to deletion of resources or more subtle errors.
